### ISSUE TITLE
omit_local_variable_types change may turn type to dynamic

### ISSUE THREAD

**Issue by awhitford** [dart-lang/linter#2154]
Consider the following code:
```dart
    final double taxableSubtotal = items
            .where((_) => true != _.menuItem.excludeFromTaxes)
            .fold(0, (sum, item) => sum + item.subtotal()) ??
        0;
```
This triggers the `omit_local_variable_types` lint rule.  However, removing the `double` then changes the type of `taxableSubtotal` from `double` to `dynamic`. 😮 

The `omit_local_variable_types` lint rule should only trigger when the removal of the type will result in the same type.

**Comment by pq** [comment]
Thanks!  I just recently patched up a similar issue w/ `unnecessary_lambdas`.  I expect a fix here would take a similar approach.

**Comment by srawlins** [comment]
I think the issue here might be specific to double/int conversion? Good bug though.
